sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -485,6 +485,10 @@ void CurlSession::ReadStatusLineAndHeadersFromRawResponse(bool reUseInternalBUff
       // Try to fill internal buffer from socket.
       // If response is smaller than buffer, we will get back the size of the response
       bufferSize = ReadFromSocket(this->m_readBuffer, Details::c_DefaultLibcurlReaderSize);
+      if (bufferSize == 0) {
+        // closed connection, prevent application from keep trying to pull more bytes from the wire
+        throw TransportException("Connection was closed by the server while trying to read a response");
```

Comment:
**Question:** If the connection is closed/no more data is left to read (because `bufferSize == 0`), should we just return and complete gracefully rather than failing? Does that mean curl is "done" reading the data or is this truly an exception case where something went wrong?

Reply:
it's a good question.
It is all related to how HTTP protocol works.
We send a request. Then the server returns a response code, headers and payload.  Usually, in the headers we would find the info about the length of the payload.
And we know how to detect where is the end of the headers.

So, we start reading until the end of headers. Here is where, if we read 0 bytes, we have to throw because we are not even at the end of headers and the connection is closed. We don't have enough data to consider the response code and headers in one useful state. (this is the current issue)

But, if we get to the end of headers, then we can parse headers to learn about the coming payload. Then we can start pulling the expected payload until the end.  At this point is easy to know if the response was not received all, because we know how much it was download vs how much we were expecting.

So, the question is, while we haven't read up to the end of the headers, there's never a scenario where there is no more data to read.